### PR TITLE
Add a method for setting token lifetimes

### LINKED ISSUE #23: Set a time limit for tokens
We should limit them to something reasonable, like 5 minutes or an hour, just so if they are ever disclosed the risk is minimal.

### PR DESCRIPTION
**Pull request by deviantintegral**
We need this API to resolve https://github.com/Lullabot/media_mpx/issues/23

### CODE REVIEW

File: tests/src/Unit/AuthenticatedClientTest.php [hunk 1]
```diff
@@ -210,6 +211,32 @@ public function testGetConfig()
         $this->assertEquals('the-value', $authenticatedClient->getConfig('the-option'));
     }
 
+    /**
+     * Test setting a duration for renewed tokens.
+     *
+     * @covers ::setTokenDuration()
```
Comment (juanolalla) [inline tests/src/Unit/AuthenticatedClientTest.php]: Is () valid syntax in @covers? I've got errors from using that

File: tests/src/Unit/AuthenticatedClientTest.php [hunk 2]
```diff
@@ -210,6 +211,32 @@ public function testGetConfig()
...
+     * @covers ::mergeAuth()
```
Comment (juanolalla) [inline tests/src/Unit/AuthenticatedClientTest.php]: I don't see how this is covering ::mergeAuth

File: tests/src/Unit/AuthenticatedClientTest.php [hunk 3]
```diff
@@ -210,6 +211,32 @@ public function testGetConfig()
...
+        $userSession = $this->getMockBuilder(UserSession::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $authenticatedClient = new AuthenticatedClient($client, $userSession);
+        $duration = rand(1, 3600);
+        $userSession->expects($this->once())->method('acquireToken')
+            ->with($duration, false)
+            ->willReturn(new Token('mpx/USER-ID', 'abcdef', $duration));
+        $authenticatedClient->setTokenDuration($duration);
+        $authenticatedClient->request('GET', 'https://www.example.com/');
```
Comment (juanolalla) [inline tests/src/Unit/AuthenticatedClientTest.php:237]: I don't see how this test works, no assertions?

**Review by juanolalla — COMMENTED** [review]
I've left some comments/questions

File: tests/src/Unit/AuthenticatedClientTest.php (hunk 1 above)
Comment (deviantintegral) [reply tests/src/Unit/AuthenticatedClientTest.php]: Interesting, what errors? This is what phpstorm is autocompleting for me, and I haven't seen any errors come up.

File: tests/src/Unit/AuthenticatedClientTest.php (hunk 3 above)
Comment (deviantintegral) [reply tests/src/Unit/AuthenticatedClientTest.php:237]: Line 233 has the assertion through the `expects()` call. If the method is called a different number of times, or with different parameters, an assertion will fail. You can try this locally by reverting the change in the `acquireToken` method call.

File: tests/src/Unit/AuthenticatedClientTest.php (hunk 2 above)
Comment (deviantintegral) [reply tests/src/Unit/AuthenticatedClientTest.php]: As mergeAuth is private, it's covered through the public method call on `request`. If you generate code coverage for this test you should see mergeAuth() being called.

File: tests/src/Unit/AuthenticatedClientTest.php (hunk 1 above)
Comment (juanolalla) [reply tests/src/Unit/AuthenticatedClientTest.php]: covers-invalid-syntax

File: tests/src/Unit/AuthenticatedClientTest.php (hunk 2 above)
Comment (juanolalla) [reply tests/src/Unit/AuthenticatedClientTest.php]: :+1:

File: tests/src/Unit/AuthenticatedClientTest.php (hunk 3 above)
Comment (juanolalla) [reply tests/src/Unit/AuthenticatedClientTest.php:237]: :+1:

File: tests/src/Unit/AuthenticatedClientTest.php (hunk 1 above)
Comment (deviantintegral) [reply tests/src/Unit/AuthenticatedClientTest.php]: This is interesting. PHPStorm adds the `()` on autocomplete, but I can't find any documentation as to why in YouTrack. The failure you're seeing is added at `\Drupal\Tests\Listeners\DrupalStandardsListenerTrait::checkValidCoversForTest` by Drupal core, which is why there's no error in this non-Drupal library. [Remove () from a bunch of @covers definitions in PHPUnit](https://www.drupal.org/project/drupal/issues/2328919) is the closest issue I can find, which boils down to "lets follow the docs".

I filed [a bug in PHPStorm](https://youtrack.jetbrains.com/issue/WI-42254) for this.

Comment (juanolalla) [reply tests/src/Unit/AuthenticatedClientTest.php]: Nice research, and great job overall!

**Pull request merged**